File: 131044021_HW8/BigramMap.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStreamReader;
import static java.lang.System.exit;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;


/**
 *
 * @author busra
 * @param <T>
 */
public class BigramMap<T> implements Bigram{
    
    protected int type;
    protected Map<Object, T> dataMap = new HashMap();

    /**
     * 
     */
    public BigramMap() {
     //empty
    }
    
    /**
     * 
     * @param dType 
     */
    public BigramMap(int dType) {
        type= dType;
    }

    /**
     * 
     * @param fName 
     */
    public void readFile(String fName) {
        
        
        FileInputStream in = null;
        try {
            in = new FileInputStream(fName);
        } catch (FileNotFoundException ex) {
            Logger.getLogger(BigramMap.class.getName()).log(Level.SEVERE, null, ex);
        }
        BufferedReader br = new BufferedReader(new InputStreamReader(in));

        String  tempLine;
        String  line="";
        String[] l=null;
        
        try {
            while((tempLine = br.readLine()) !=null)
            {
                line += tempLine;
                
            }
            
            l =line.split("\\s+");
      
            for (int j = 0; j < l.length; j++){
                dataMap. put(j, (T)l[j]);
                //System.out.print(" " + dataMap.get(j));
            }
            
        } catch (IOException ex) {
            Logger.getLogger(BigramMap.class.getName()).log(Level.SEVERE, null, ex);
        }
        
        
        try {
            in.close();
                   
        } catch (IOException ex) {
            Logger.getLogger(BigramDyn.class.getName()).log(Level.SEVERE, null, ex);
        }
    }

    /**
     * 
     * @return 
     */
    public int numGrams() {
 
        return dataMap.size()- 1;    
    }

    /**
     * 
     * @param p1
     * @param p2
     * @return 
     */
    public int numOfGrams(Object p1, Object p2) {
        int num=0;
        List<Integer> dataInt = new ArrayList<>();
        List<Double> dataDouble= new ArrayList<>();
        List<String> dataString = new ArrayList<>();
        
        String message= "Tip ayni degildir.Tum tipler string olarak dusunulerek bigram sayisi bastirilmistir\n";
        
        try{
            if(type==1)
            {
                for(int j=0; j<=numGrams(); ++j)
                    dataInt.add(j, Integer.parseInt((String) dataMap.get(j)));
                
                num=0;
                for(int i=0; i<= numGrams(); ++i){
                    if(i != numGrams()) { 

                       if(dataInt.get(i) == (int)p1 && dataInt.get(i+1) == (int)p2)
                            ++num;
                    }
                }
            }
             if(type==3)
            {
                for(int j=0; j<=numGrams(); ++j)
                    dataDouble.add(j, Double.parseDouble((String) dataMap.get(j)));

                num=0;
                for(int i=0; i<= numGrams(); ++i){
                    if(i != numGrams()) { 

                       if(dataDouble.get(i) == (double)p1 && dataDouble.get(i+1) == (double)p2)
                            ++num;
                    }
                }
            }
            else if(type==2)
            {
                for(int j=0; j<=numGrams(); ++j)
                    dataString.add(j, String.valueOf(dataMap.get(j)));

                num=0;
                for(int i=0; i<= numGrams(); ++i){
                    if(i != numGrams()) { 

                       if((dataString.get(i) == null ? (String)p1 == null : dataString.get(i).equals((String)p1)) && dataString.get(i+1).equals((String)p2))
                            ++num;
                    }
                }
            }  
        }catch(NumberFormatException e){
            System.out.println(message + "NumberFormatException throw" );
            exit(1);
        }
              
       return num;
     
    }

    /**
     * 
     * @return 
     */
    @Override
    public String toString() 
    {
        String result;
        result = "\nPrints All the Bigram\n";
        for(int i=0; i<numGrams(); ++i){
           result += dataMap.get(i) + " " + dataMap.get(i+1) + "\n";
        }
        
       
        //Bigram Siralamasi
        if(type == 1)  //integer
        {
            result += "\nDecreasing Sorting Bigrams\n";
            int maxS=1;
            int[] arrMaxS = new int[numGrams()];
            int k=0;

            for(int i=0; i< numGrams(); ++i){
                for(int j=i+1; j< numGrams(); ++j){
                    if(Integer.parseInt((String) dataMap.get(i)) == Integer.parseInt((String) dataMap.get(j))){
                        if(Integer.parseInt((String) dataMap.get(i+1)) == Integer.parseInt((String) dataMap.get(j+1)))
                            ++maxS;
                    }
                }  
                arrMaxS[k]=maxS; //kac kere tekrar ettiklerini tutan array
                ++k;
                maxS=1;
            } 

            //en cok tekrar eden bigramin indexini bulma
            int tempMax=0, t=0, s=0, p=0;
            int maxRepeatEle=0;

            for(t=0; t< numGrams(); ++t){

                if(arrMaxS[t] > maxRepeatEle){
                    maxRepeatEle= arrMaxS[t];
                    tempMax=t;       
                }
            }
           
            int[] tempInt = new int[numGrams()*2];
            int y=0;
            //maximumdan baslayacak sekilde bigramlari siralama
            for(s=maxRepeatEle; s>0; --s){
                for(p=0; p< numGrams(); ++p){
                    if(arrMaxS[p] == s){
                        tempInt[y]= Integer.parseInt((String) dataMap.get(p));
                        tempInt[y+1]= Integer.parseInt((String) dataMap.get(p+1));
                        y+=2;   
                    }
                   
                }
            }              

            int tmp=0;
            result += tempInt[0] + " " + tempInt[1] + "\n";
            for(int a=2; a< numGrams()*2; a+=2){   
                tmp=0;
                for(int x=0; x<a; x+=2){
                    if(tempInt[a]== tempInt[x] && tempInt[a+1]== tempInt[x+1]){
                        tmp=1;
                    }
                }
                if( tmp!=1){
                    result += tempInt[a] + " " + tempInt[a+1] + "\n";
                }             
            }
        }
        else if(type == 2)  //string
        {
            
            result += "\nDecreasing Sorting Bigrams\n";
            int maxS=1;
            int[] arrMaxS = new int[numGrams()];
            int k=0;

            for(int i=0; i< numGrams(); ++i){
                for(int j=i+1; j< numGrams(); ++j){
                    if(String.valueOf(dataMap.get(i)).equals(String.valueOf(dataMap.get(j)))){
                        if(String.valueOf(dataMap.get(i+1)).equals(String.valueOf(dataMap.get(j+1))))
                            ++maxS;
                    }
                }  
                arrMaxS[k]=maxS; //kac kere tekrar ettiklerini tutan array
                ++k;
                maxS=1;
            } 

            //en cok tekrar eden bigramin indexini bulma
            int tempMax=0, t=0, s=0, p=0;
            int maxRepeatEle=0;

            for(t=0; t< numGrams(); ++t){

                if(arrMaxS[t] > maxRepeatEle){
                    maxRepeatEle= arrMaxS[t];
                    tempMax=t;       
                }
            }
            int y=0;
            String[] tempString = new String[numGrams()*2];
            for(s=maxRepeatEle; s>0; --s){
                for(p=0; p< numGrams(); ++p){
                    if(arrMaxS[p] == s){
                        tempString[y]= String.valueOf(dataMap.get(p));
                        tempString[y+1]= String.valueOf(dataMap.get(p+1));
                        y+=2;    
                    }
                }
            }   

            int tmp=0;
            result += tempString[0] + " " + tempString[1] + "\n";
            for(int a=2; a< numGrams()*2; a+=2){   
                tmp=0;
                for(int x=0; x<a; x+=2){
                    if(tempString[a].equals(tempString[x]) && tempString[a+1].equals(tempString[x+1])){
                        tmp=1;
                    }
                }
                if( tmp!=1){
                    result += tempString[a] + " " + tempString[a+1] + "\n";
                }             
            }
        }
        else if(type == 3)   //double
        {
            result += "\nDecreasing Sorting Bigrams\n";
            int maxS=1;
            int[] arrMaxS = new int[numGrams()];
            int k=0;

            for(int i=0; i< numGrams(); ++i){
                for(int j=i+1; j< numGrams(); ++j){
                    if(Double.parseDouble((String) dataMap.get(i)) == Double.parseDouble((String) dataMap.get(j))){
                        if(Double.parseDouble((String) dataMap.get(i+1)) == Double.parseDouble((String) dataMap.get(j+1)))
                            ++maxS;
                    }
                }  
                arrMaxS[k]=maxS; //kac kere tekrar ettiklerini tutan array
                ++k;
                maxS=1;
            } 

            //en cok tekrar eden bigramin indexini bulma
            int tempMax=0, t=0, s=0, p=0;
            int maxRepeatEle=0;

            for(t=0; t< numGrams(); ++t){

                if(arrMaxS[t] > maxRepeatEle){
                    maxRepeatEle= arrMaxS[t];
                    tempMax=t;       
                }
            }


            int y=0;  
            double[] tempDouble = new double[numGrams()*2];
            
            for(s=maxRepeatEle; s>0; --s){
                for(p=0; p< numGrams(); ++p){
                    if(arrMaxS[p] == s){
                        tempDouble[y]= Double.parseDouble((String) dataMap.get(p));
                        tempDouble[y+1]= Double.parseDouble((String) dataMap.get(p+1));
                        y+=2;    
                    }
                }
            }   

            int tmp=0;
            result += tempDouble[0] + " " + tempDouble[1] + "\n";
            for(int a=2; a< numGrams()*2; a+=2){   
                tmp=0;
                for(int x=0; x<a; x+=2){
                    if(tempDouble[a]== tempDouble[x] && tempDouble[a+1]== tempDouble[x+1]){
                        tmp=1;
                    }
                }
                if( tmp!=1){
                    result += tempDouble[a] + " " + tempDouble[a+1] + "\n";
                }             
            }
        } 
             
       return result;
    } 
}

File: 131044021_HW7/BigramMap.cpp
#include <iostream>
#include <stdlib.h> 
#include <fstream>  //dosya okuma kutuphanesi 
#include <string>   //cpp string kutuphanesi
#include <cstring> 
#include <vector>
#include <map>     //stl map class
#include <utility>

#include "Bigram.h"
#include "BigramMap.h" 

using namespace std;

namespace Busra{
    
    template<class T>
    BigramMap<T>::BigramMap()
    { /*empty*/ }
    
    template<class T>
    BigramMap<T>::BigramMap(int  dataType)
    {
        try{
             string msg= "Datatype negatif deger olamaz. 1,2 ve 3 olabilir";
             if(dataType==1 || dataType==2 || dataType==3)
				type= dataType;
             else
                throw msg; 
        }catch (string e){
			cout << e << " Exception throw"  << endl;
                //exit(1);
		}
    }
    
    
    template<class T>
    void BigramMap<T>::readFile(const char* fName)
    {
        T tempData;
        int temp2=0;
        string message= "Don't read file";
        int tempSize=0, emptyControl=0;
        string message_= "Veriler ayni tip degildir!";
        string s;
        listSize=0;
        
        if(type == 2)
            cout << "Datatype string olarak secilmistir.integer ve double degerlerde stringdir." << endl;
        
        //exception handling dosya okuma hata durumu
        try{
            ifstream inputFile;
            inputFile.open(fName);

            if(inputFile)  //hatali ise exception firlat
            {          
                int i=0;
                while(inputFile>>tempData)
                {
                    dataList[i]=tempData;
                    //cout << dataList[i] << " ";    
                    ++i;
                }  
                tempSize=i;
                listSize=i;
                inputFile.close();
                inputFile.open(fName);
            }else if(!inputFile)
                throw message;
            if (inputFile.is_open()) {
                while (inputFile >> s  ) {   //string okumasi
                        ++temp2;
                }
                inputFile.close();
            }else if(!inputFile)
                throw message; 
    
        
        }catch (string e){
		cout << e << " Exception throw"  << endl;
                exit(1);
	}
        
        /////dosya bos mu kontrolu////
        
        string message2="File is empty";
        try{
            if(emptyControl == numGrams()+1 && temp2==0)
                throw message2;
        }catch(string e){
                cout << e << " Exception throw"  << endl;
                exit(1);
        }
  
     
        //dosyadaki T tipi verilerin hepsi ayni tip mi kontrolu
        try{
            if( temp2 != tempSize)
                throw message_;
            
        }catch(string e){
            cout<< e << "Exception throw" << endl;
            if(type == 1)
                cout << tempSize << ". eleman farkli tiptedir." << endl;
            else if(type==3)
                cout << tempSize << ". eleman farkli tiptedir." << endl;
            exit(1);
        }
        
        for(int i=0; i< tempSize; ++i)
            cout << dataList[i] << " ";
        cout<< endl;
    }
    
    
    template<class T>
    int BigramMap<T>::numGrams()
    {
        string msg="size sifirdan kucuk deger olamaz";

        try{
            if(dataList.size() >=0) 
                return dataList.size()-1;
           
            else
                throw msg;
        }catch(string e){
            cout << e << "Exception throw"<< endl;
        }
        
    }
    
    //sirali ikililerden kac tane varsa o kadar dondurur
    template<class T>
    int BigramMap<T>::numOfGrams(T& p1, T& p2)
    {
        string msg="Sifirdan kucuk deger olamaz numOfGrams";
        int num=0;
        for(int i=0; i<= numGrams(); ++i){
            if(i != numGrams()) { 
                if(dataList.at(i) == p1 && dataList.at(i+1) == p2)
                    ++num;
            }
        }
        
        try{
            if(num >=0) 
                return num;
           
            else
                throw msg;
        }catch(string e){
            cout << e << "Exception throw"<< endl;
        }
         
    }
    
    
    template<class T>
    std::pair<T, T> BigramMap<T>::maxGrams()
    {
        pair<T, T> p;
        
        int maxS=1;
        int arrMaxS[numGrams()];
        int k=0;
        
        for(int i=0; i<= numGrams(); ++i){
            for(int j=i+1; j< numGrams(); ++j){
                if(dataList.at(i) == dataList.at(j)){
                    if(dataList.at(i+1) == dataList.at(j+1))
                        ++maxS;
                }
            }  
            arrMaxS[k]=maxS;
            ++k;
            maxS=1;
        }
        
        //en cok tekrar eden bigramin indexini bulma
        int tempMax=0, t=0;
        int maxRepeatEle=0;
        
        for(t=0; t< numGrams(); ++t){
            if(arrMaxS[t] > maxRepeatEle){
                maxRepeatEle= arrMaxS[t];
                tempMax=t;
            }
        }
        
        //cout<< myData[tempMax] <<" " << myData[tempMax+1] << endl;

        p=std::make_pair(dataList[tempMax], dataList[tempMax+1]);
        
        return p;
    }
 
    
    template<class T>
    void BigramMap<T>::print(ostream & out)const
    {
        out<< "Prints All the Bigram" << endl;
        
       for(int i=0; i<listSize-1; ++i)
           out<< dataList.at(i) << " " << dataList.at(i+1)<< endl;
        
        cout << endl;
        
        out<< "Decreasing Sorting Bigrams" << endl;
        int maxS=1;
        int arrMaxS[listSize-1];
        int k=0;
        
        for(int i=0; i< listSize-1; ++i){
            for(int j=i+1; j< listSize-1; ++j){
                if(dataList.at(i) == dataList.at(j)){
                    if(dataList.at(i+1) == dataList.at(j+1))
                        ++maxS;
                }
            }  
            arrMaxS[k]=maxS; //kac kere tekrar ettiklerini tutan array
            ++k;
            maxS=1;
        } 

        //en cok tekrar eden bigramin indexini bulma
        int tempMax=0, t=0, s=0, p=0;
        int maxRepeatEle=0;
        
        for(t=0; t< listSize-1; ++t){
            if(arrMaxS[t] > maxRepeatEle){
                maxRepeatEle= arrMaxS[t];
                tempMax=t;
               
            }
        }

        int y=0,z=0;
        int bigramSize=2;
        T tempList[listSize-1][bigramSize];
        for(s=maxRepeatEle; s>0; --s){
            for(p=0; p< listSize-1; ++p){
                if(arrMaxS[p] == s)
                { 
                    tempList[y][z] = dataList.at(p);
                    tempList[y][z+1] = dataList.at(p+1);
                    ++y;  z=0;   
                }
            }
        }    
//        int b=0;
//        for(int a=0; a< (dataSize-1); ++a)
//      
//                out<< tempList[a][b]<< " " << tempList[a][b+1] << endl;
        
        int tmp=0;
        out<< tempList[0][0]<< " " << tempList[0][1] << endl; 
        for(int a=1; a< (listSize-1); ++a){
            int b=0;   
            tmp=0;
            for(int x=0; x<a; ++x){
                if(tempList[a][b] == tempList[x][b] && tempList[a][b+1] == tempList[x][b+1]  ){
                    tmp=1;
                }
            }
            if( tmp!=1){
                out<< tempList[a][b]<< " " << tempList[a][b+1] << endl;
            }             
        }
        
    }
    
}// end namespace


